### DISCUSSION TITLE
Storybook autodocs feature in Vue files

### DISCUSSION THREAD

**Discussion by RomanOnTheHill** [storybookjs/storybook#24230] [Help]
I have a Vue 3 component in my Storybook 7.4 project. We use a composite API with "setup", like `<script lang="ts" setup>`. I want to use the "autodocs" feature described in the [Storybook doc](https://storybook.js.org/docs/vue/writing-docs/autodocs). Unfortunately, there is no example how to do it with the "setup" option there. Here is an example file:

```vue
<template>
  <q-btn @click="emit('click')">
    <template v-slot:default>
      <div v-if="props.label">{{ props.label }}</div>
    </template>
  </q-btn>
</template>

<script lang="ts" setup>
const props = defineProps({
  /**
   * This is the only working autodocs.
   */
  label: {
    type: String,
    required: false,
  },
});

/**
 * It doesn't appear anywhere.
 */
const emit = defineEmits(['click']);
</script>

<style></style>
```
The only working JDoc is in the `defineProps`. I've tried adding comments here and there, but to no avail. Is there a way to use the autodocs feature in such files? Or any other suitable way? I don't want to write MDX, it seems to me very complex.

**Comment by MagnusHodne** [comment]
Pro tip: if you're on Vue 3.3+, you can use the `defineOptions()` macro to generate a top-level description (same as if you were to add a JSDoc comment above the `export default` in an Options API component). In the case of defineEmits, you have to place the comment above each entry in your array, not at the top of the defineEmits

The autodocs generated for script setup components still isn't quite as good as for options API (using a string union type for a prop in options API yields a radio group control in Storybook, but with script setup we only get a string input field). I think this is an issue with vue-docgen-api (which is what Storybook uses to parse documentation in SFCs) and not Storybook though...

```vue
<script setup lang="ts">
/** Some top-level documentation describing your component here */
defineOptions()

defineEmits([
  /** Document your click event here */
  'click'
])
</script>
```

https://github.com/vue-styleguidist/vue-styleguidist/blob/dev/packages/vue-docgen-api/src/script-setup-handlers/setupOptionsHandler.test.ts Based on the test written for vue-docgen-api it looks like this is the intended behaviour for the defineOptions macro as well, so even though this ends up looking like a hack, I actually think this might be the way you're "supposed" to use autodocs with script setup (but don't quote me on that) 😅
